src/platform/i386/chal_pgtbl.c

```diff
@@ -0,0 +1,533 @@
+/**
+ * Copyright 2018 by Runyu Pan, panrunyu@gwu.edu
+ *
+ * Redistribution of this file is permitted under the GNU General
+ * Public License v2.
+ */
+#include "shared/cos_types.h"
+#include "captbl.h"
+#include "pgtbl.h"
+#include "cap_ops.h"
+#include "liveness_tbl.h"
+#include "retype_tbl.h"
+#include "chal/chal_prot.h"
+#include "chal_pgtbl.h"
+
+/* These functions do flag operations */
+unsigned long
+chal_pgtbl_flag_add(unsigned long input, pgtbl_flags_t flags)
+{
+	return input | flags;
+}
+
+unsigned long
+chal_pgtbl_flag_clr(unsigned long input, pgtbl_flags_t flags)
+{
+	return input & (~flags);
+}
+
+unsigned long
+chal_pgtbl_flag_exist(unsigned long input, pgtbl_flags_t flags)
+{
+	return input & flags;
+}
+
+unsigned long
+chal_pgtbl_flag_all(unsigned long input, pgtbl_flags_t flags)
+{
+	return chal_pgtbl_flag_exist(input, flags) == flags;
+}
+
+int
+chal_pgtbl_kmem_act(pgtbl_t pt, u32_t addr, unsigned long *kern_addr, unsigned long **pte_ret)
+{
+	struct ert_intern *pte;
+	u32_t              orig_v, new_v, accum = 0;
+
+	assert(pt);
+	assert((PGTBL_FLAG_MASK & addr) == 0);
+
+	/* get the pte */
+	pte = (struct ert_intern *)__pgtbl_lkupan((pgtbl_t)((u32_t)pt | X86_PRESENT), addr >> PGTBL_PAGEIDX_SHIFT,
+	                                          PGTBL_DEPTH, &accum);
+	if (unlikely(!pte)) return -ENOENT;
+	if (unlikely(__pgtbl_isnull(pte, 0, 0))) return -ENOENT;
+
+	orig_v = (u32_t)(pte->next);
+	if (unlikely(!(orig_v & X86_COSFRAME))) return -EINVAL; /* can't activate non-frames */
+	if (unlikely(orig_v & X86_COSKMEM)) return -EEXIST;     /* can't re-activate kmem frames */
+	assert(!(orig_v & X86_QUIESCENCE));
+
+	*kern_addr = (unsigned long)chal_pa2va((paddr_t)(orig_v & PGTBL_FRAME_MASK));
+	new_v      = orig_v | X86_COSKMEM;
+
+	/* pa2va (value in *kern_addr) will return NULL if the page is not kernel accessible */
+	if (unlikely(!*kern_addr)) return -EINVAL; /* cannot retype a non-kernel accessible page */
+	if (unlikely(retypetbl_kern_ref((void *)(new_v & PGTBL_FRAME_MASK)))) return -EFAULT;
+
+	/**
+         * We keep the cos_frame entry, but mark it as COSKMEM so that
+	 * we won't use it for other kernel objects.
+         */
+	if (unlikely(cos_cas((unsigned long *)pte, orig_v, new_v) != CAS_SUCCESS)) {
+		/* restore the ref cnt. */
+		retypetbl_deref((void *)(orig_v & PGTBL_FRAME_MASK));
+		return -ECASFAIL;
+	}
+	/**
+         * Return the pte ptr, so that we can release the page if the
+	 * kobj activation failed later.
+         */
+	*pte_ret = (unsigned long *)pte;
+
+	return 0;
+}
+
+int
+chal_tlb_quiescence_check(u64_t timestamp)
+{
+	int i, quiescent = 1;
+
+	/**
+         * Did timer interrupt (which does tlb flush
+	 * periodically) happen after unmap? The periodic
+	 * flush happens on all cpus, thus only need to check
+	 * the time stamp of the current core for that case
+	 * (assuming consistent time stamp counters).
+         */
```

Comment:
spacing off before `*` in comments.

Reply:
no reason to have a `/**` comment *in* a function.

Your spacing is off here.   Likely another space/tab problem.  Please take care of these in your editor.  I don't want to have to look this carefully at every line of code.